lib/TH/THTensorApply.h

```diff
@@ -34,70 +36,131 @@
   else \
   { \
     TENSOR1##_data = TENSOR1->storage->data+TENSOR1->storageOffset; \
-    for(TENSOR1##_dim = TENSOR1->nDimension-1; TENSOR1##_dim >= 0; TENSOR1##_dim--) \
-    { \
-      if(TENSOR1->size[TENSOR1##_dim] != 1) \
-        break; \
-    } \
-    TENSOR1##_stride = (TENSOR1##_dim == -1 ? 0 : TENSOR1->stride[TENSOR1##_dim]); \
     TENSOR1##_size = 1; \
-    for(TENSOR1##_dim = TENSOR1->nDimension-1; TENSOR1##_dim >= 0; TENSOR1##_dim--) \
-    { \
-      if(TENSOR1->size[TENSOR1##_dim] != 1) \
+    TENSOR1##_stride = 1; \
+    for(TENSOR1##_i = TENSOR1->nDimension-1; TENSOR1##_i >= 0; TENSOR1##_i--) { \
+      if(TENSOR1->size[TENSOR1##_i] != 1) { \
+        if(TENSOR1->stride[TENSOR1##_i] == TENSOR1##_size) \
+          TENSOR1##_size *= TENSOR1->size[TENSOR1##_i]; \
+        else{ \
+          TH_TENSOR1_contiguous = 0; \
+	  break; \
+	} \
+      } \
+    } \
+    if (!TH_TENSOR1_contiguous) { \
```

Comment:
Wouldn't it be better to separate these parts into a different macro and just instantiate it 3 times in here? There seems to be a lot of redundancy.